pkg/internalregistry/controllers/service_account_controller.go

```diff
@@ -134,16 +139,56 @@ func (c *serviceAccountController) sync(ctx context.Context, key string) error {
+	patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+
+	// we always want the annotation set
+	patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
+
+	// TODO stop adding to .Secrets part of API-1798
+	if !slices.ContainsFunc(patch.Secrets, func(ref applycorev1.ObjectReferenceApplyConfiguration) bool { return *ref.Name == secretName }) {
+		patch.WithSecrets(applycorev1.ObjectReference().WithName(secretName))
+	}
+
+	// We have tp be careful with the atomic ServiceAccount.ImagePullSecrets field so
+	// that we do not race with another controller or user. Add ImagePullSecrets to
+	// the apply patch explicitly only if the value we want to set it to was not the
+	// value extracted by the ExtractServiceAccount() func, and if the managed image
+	// pull secrets is not already referenced in the service account ImagePullSecrets
+	// field.
+	//  | Contains managed image pull secret ref:
+	//  |               \ |                   patch                             |
+	//  | serviceaccount \|  true               | false                         |
+	//  | --------------- | ------------------- | ----------------------------- |
+	//  | true            |  Already in patch   | Ignore (other owner exists)   |
+	//  | false           |  Should not happen  | Add to patch (take ownership) |
+	//
+	serviceAccountContainsImagePullSecretRef := slices.ContainsFunc(serviceAccount.ImagePullSecrets, func(ref corev1.LocalObjectReference) bool { return ref.Name == secretName })
+	patchContainsImagePullSecretRef := slices.ContainsFunc(patch.ImagePullSecrets, func(ref applycorev1.LocalObjectReferenceApplyConfiguration) bool { return *ref.Name == secretName })
+	if !serviceAccountContainsImagePullSecretRef && !patchContainsImagePullSecretRef {
+		patch.ImagePullSecrets = nil
+		for _, ref := range serviceAccount.ImagePullSecrets {
+			patch.WithImagePullSecrets(applycorev1.LocalObjectReference().WithName(ref.Name))
 		}
-		patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
+		patch.WithImagePullSecrets(applycorev1.LocalObjectReference().WithName(secretName))
+	}
+
+	// apply patch only if necessary
+	extracted, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+	if !cmp.Equal(patch, extracted) {
 		// add the UID to the patch to ensure we don't re-create the service account if it no longer exists
 		patch.WithUID(serviceAccount.UID)
-		// we apply this now to ensure the secret name stays stable in case a error occurs while reconciling
+		// if setting the atomic ImagePullSecrets field, add the resourceVersion to the
+		// patch to ensure we do not race with another operator or user.
+		if len(patch.ImagePullSecrets) > 0 {
+			patch.WithResourceVersion(serviceAccount.ResourceVersion)
+		}
```

Comment:
If we're building the desired configuration from scratch, then ResourceVersion could be set at the same time ImagePullSecrets.

Reply:
I refactored to start with a empty patch instead of an extracted one, but it still make sense to me to add the RV after the desired/extracted patch comparison.